RecoMuon/TrackerSeedGenerator/plugins/TSGForOI.cc

```diff
@@ -78,298 +94,301 @@ void TSGForOI::produce(edm::StreamID sid, edm::Event& iEvent, const edm::EventSe
+    // If both TSOSs agree, use only the one at vertex, as it uses more information. If they do not agree, search for seeds based on both.
     bool useBoth = false;
-    if (outerTkStateInside.isValid() && outerTkStateOutside.isValid()){
-	dist = match_Chi2(outerTkStateInside,outerTkStateOutside);
-    }
-    if (dist>tsosDiff_){
-	useBoth = true; 
+    if (outerTkStateInside.isValid() && outerTkStateOutside.isValid()) {
+      //auto dist1 = match_Chi2(outerTkStateInside,outerTkStateOutside);
+      //auto dist2 = deltaR(outerTkStateInside.globalMomentum(),outerTkStateOutside.globalMomentum());
+      //if ((dist1 > tsosDiff1_ || dist2 > tsosDiff2_) && l2->numberOfValidHits() < 20) useBoth = true;
+      if (l2->numberOfValidHits() < 20) useBoth = true;
+      if (l2->numberOfValidHits() < 30 && (std::abs(l2->eta()) > 2.1)) useBoth = true;
+      //if (std::abs(l2->eta())<1.0)useBoth=false;
+      if (std::abs(l2->eta()) > 0.2 && std::abs(l2->eta()) < 0.3) useBoth = true;
     }
 
-    numSeedsMade=0;
-    analysedL2 = false;
-
-    // if both TSOSs agree, use only the one at vertex, as it uses more information. If they do not agree, search for seed based on both
-
-    //		BARREL
+    numSeedsMade = 0;
+    hitlessSeedsMadeIP = 0;
+    hitlessSeedsMadeMuS = 0;
+    hitSeedsMade = 0;
+
+    // calculate scale factors
+    double errorSFHits = 1.0;
+    double errorSFHitless = 1.0;
+    if (!adjustErrorsDynamicallyForHits_) errorSFHits = fixedErrorRescalingForHits_;
+    else                                  errorSFHits = calculateSFFromL2(l2);
+    if (!adjustErrorsDynamicallyForHitless_) errorSFHitless = fixedErrorRescalingForHitless_;
+    else                                  errorSFHitless = calculateSFFromL2(l2);
+      
+    // BARREL
     if (std::abs(l2->eta()) < maxEtaForTOB_) {
       layerCount = 0;
-      for (auto it=tob.rbegin(); it!=tob.rend(); ++it) {	//This goes from outermost to innermost layer
-	LogTrace("TSGForOI") << "TSGForOI::produce: looping in TOB layer " << layerCount << endl; 
-	findSeedsOnLayer(tTopo, **it, tsosAtIP,  *(propagatorAlong.get()), *(propagatorOpposite.get()), l2, 
-			 estimatorH, measurementTrackerH, numSeedsMade, numOfMaxSeeds, layerCount, analysedL2, out);
+      for (auto it=tob.rbegin(); it!=tob.rend(); ++it) { 
+        LogTrace("TSGForOI") << "TSGForOI::produce: looping in TOB layer " << layerCount << endl;
+        if ( useHitLessSeeds_ && hitlessSeedsMadeIP < maxHitlessSeeds_ && numSeedsMade < maxSeeds_ ) makeSeedsWithoutHits(**it, tsosAtIP, l2, *(propagatorAlong.get()), estimatorH, errorSFHitless, hitlessSeedsMadeIP, numSeedsMade, out);
+        // Do not create hitbased seeds in barrel region
+        if (std::abs(l2->eta()) > 1.0 && hitSeedsMade < maxHitSeeds_ && numSeedsMade < maxSeeds_ ) makeSeedsFromHits(**it, tsosAtIP, l2, *(propagatorAlong.get()), estimatorH, measurementTrackerH, errorSFHits, hitSeedsMade, numSeedsMade, layerCount, out);
```

Comment:
please add line breaks to improve readability

Reply:
Okay, Will be added